### PR TITLE
Ensure that `reactiveValues` keys and values are sorted

### PR DESCRIPTION
**Pull request by wch**
This PR also ensures that anywhere that a `reactiveValues` `$names()` or `$values()` method is called, the keys will be sorted in the same order that they were inserted. (Note: a previous version of this PR sorted lexicographically, and some of the comments were for that version.)

In our shinycoreci tests, we found that bookmarked input values were not always in consistent order. For example, for the same app in the same state, both of these URLs might be generated:

* http://127.0.0.1/?_inputs_&reset=0&x=5
* http://127.0.0.1/?_inputs_&x=5&reset=0

### CODE REVIEW

**Comment by jcheng5** [comment]
I'm OK with this but it's a slight shame that they're not in insertion order, like a named list. @rpodcast and I needed this the other day and ended up resorting to tracking the insertion order in a side list.

**Comment by wch** [comment]
Just curious - what did you need insertion order for?

**Comment by rpodcast** [comment]
> Just curious - what did you need insertion order for?

@jcheng5 and I have been working on improving my workflow of dynamic UI insertions into a Shiny app based on some real life examples. The goal is to have a custom R6 class to manage a lot of the details for an improved user experience over my existing methods. In some of my applications, the order of how these dynamic UIs are entered actually matters as they can represent sequential parameter sets that need to follow the order of insertion. I realize that's not a very common use case of dynamic UI, but for my purposes it can be very helpful.

**Comment by wch** [comment]
@jcheng5 I'll update this PR to return keys in insertion order.

File: tests/testthat/test-reactivity.r [hunk 1]
```diff
@@ -126,6 +126,23 @@ test_that("ReactiveValues", {
...
 
+test_that("reactiveValues keys are sorted", {
+  values <- reactiveValues(b=2, a=0)
+  values$C <- 13
+  values$A <- 0
+  values$c <- 3
+  values$B <- 12
+  # Setting an existing value shouldn't change order
+  values$a <- 1
+  values$A <- 11
```
Comment (jcheng5) [inline tests/testthat/test-reactivity.r:137]: I guess we don't have the notion of removing a value? That's the only other case I could think of.

**Review by jcheng5 — APPROVED** [review]

File: tests/testthat/test-reactivity.r (hunk 1 above)
Comment (wch) [reply tests/testthat/test-reactivity.r:137]: Right, we don't have a way to remove a value.

**Comment by daattali** [comment]
I've ran into this issue many times as well, I'll see if my 4 year old issue can be closed https://github.com/rstudio/shiny/issues/2629

**Comment by daattali** [comment]
Yep, I love when you guys fix old issues I've been waiting on for years 🥳

**Pull request merged**